### DISCUSSION TITLE
mesh triangulation of extrude with nDivisions skips triangles

### DISCUSSION THREAD

**Discussion by Loosetooth** [elalish/manifold#413] [Q&A] [answered]
Using `extrude` with a parameter `nDivisions` set greater than 1 results in a linear extrusion with some 'missing' triangles. To be clear here, the mesh is fine for the linear extrusion, but the missing triangles become apparent after trying a warp function.

I found this by trying to write a `warpFunc` to do a partial revolve of a polygon. Try out the following example code in manifoldcad:

```js

const generateProfile = () => {
  const arcCenterPoint = [
    2.5,
    2.5
  ]
  const arcPoints = [...Array(7).keys()]
    .map(i => {
      const angle = i * Math.PI / (7)
      const x = Math.cos(angle) * 2.5
      const y = Math.sin(angle) * 2.5
      return [arcCenterPoint[0] + y, arcCenterPoint[1] - x]
    })

  const profile = [
    [0, 0],
    [2.5, 0],
    ...arcPoints,
    [2, 5],
    [2, 18],
    [0, 20]
  ]

  return profile
}

const profile = generateProfile()

const partialRevolve = (startAngle, endAngle, nSegments) => {
  const totalAngle = Math.abs(endAngle - startAngle)
  const angleStep = totalAngle / (nSegments - 1)

  const warpFunc = (vertex) => {
    // slice index is encoded z height
    const zIndex = nSegments - 1 - vertex[2]
    const angle = zIndex * angleStep + startAngle
    // do transform
    vertex[2] = vertex[1]
    vertex[1] = vertex[0] * Math.sin(angle)
    vertex[0] = vertex[0] * Math.cos(angle)
  }

  return extrude(profile, nSegments - 1, nSegments - 2)
    .warp(warpFunc)
}

const result = compose([
  extrude(profile, 6 - 1, 6 - 2)
    .rotate([90,0,0])
    .translate([0, 10, 0]),
  partialRevolve(-Math.PI / 4, Math.PI / 2, 6)
])

```

This results in the following:

![image](https://user-images.githubusercontent.com/66420755/232448273-55cba794-7aeb-4fb2-b9e3-944582e77b12.png)

As you can see in the partially revolved shape, at the top and bottom of the arc segment, all the triangles are there. While in the middle of the arc, seemingly `nDivisions` did not generate any extra vertices.

I'm aware that I can use `revolve` with a warp instead to achieve my intended result, but I still wanted to share this, just to check if this is intended behavior. If so, feel free to close this issue.

**Comment by pca006132** [comment]
I'm wondering if the vertices are removed in the simplification pass as they contribute nothing to the mesh (before warp). I think `nDivisions` was designed to work with non-zero `twistDegrees`.

**Reply by elalish** [reply]
Hmm, I was going to say this is expected behavior, but I just realized I'm a bit surprised our simplify code is being called during extrude or warp. I'll look into that. 

I'd recommend a `revolve` followed by an intersection with a union or intersection of two rotated cubes to get a partial revolution. And if you have trouble with a `warp` function's granularity, you can try `refine` first.

**Reply by Loosetooth** [reply]
> I'd recommend a revolve followed by an intersection with a union or intersection of two rotated cubes to get a partial revolution. And if you have trouble with a warp function's granularity, you can try refine first.

The problem with this approach is indeed the granularity of the warp/extrude. I would like the two faces of the partial revolution -start and end- to be exactly the same as the profile (polygon) that's being extruded.

If I could somehow prevent the simplification pass after the linear extrusion, the end result should be exactly what I'm aiming for here. Using refine would add a lot of extra triangles, and thus also a lot of extra compute required in the remaining code, which uses a multiple of these extrusions. So I'd prefer to avoid using refine.

**Reply by Loosetooth** [reply]
Now that I think about this more: skipping the simplification step would also allow 'sweep' extrusions using 'warp' that are much more simple. 

Right now you would have to do a `revolve`->`warp` combination.

Instead of having to calculate what 'slice' a vertex is in using `Math.atan2` (calculating the vertex angle around z, following the revolve), the slice can be encoded in the z-height of the linear extrusion!

So we could do `extrude`->`warp`, where the z coordinate of the vertex can be used to directly figure out the slice index. This would make a future `extrudeFromSlices` function (similar to the JSCAD implementation) or `sweep` implementation trivial to implement.

**Reply by Loosetooth** [reply]
On top of the speed and simplification advantages, using extrude->warp would also offer the following advantage out of the box:

* warped mesh has end caps (i.e. a partial revolution is closed off, currently impossible using revolve)

**Reply by jirihon** [reply]
I believe all the pieces needed for implementing `extrudeFromSlices` from JSCAD are already provided by manifold. Just triangulate the cap slices (using the `triangulate` function) and calculate triangles connecting the slices. And then pass all the triangles to the `Mesh` constructor to make a `Manifold`.

**Reply by jirihon** [reply]
Basic `sweep` implementation is also possible using the same approach. I've already tried something like that. And if you care about visualization, you can also calculate the normals as you sweep and have nice smooth look. And manifold will keep your normals correct throughout the boolean operations! JSCAD can't do that.

**Reply by Loosetooth** [reply]
@jirihon 
Thanks for tuning in!

> I believe all the pieces needed for implementing extrudeFromSlices from JSCAD are already provided by manifold.

OK, would you like to share some code examples? Perhaps something that runs in manifoldcad would be nice.

> Just triangulate the cap slices (using the triangulate function)

I haven't been able to get the `triangulate` JS binding working in manifoldcad. It can't find the function. 
`ReferenceError: triangulate is not defined`
Could something be broken with the export? Or perhaps manifold.org is not on the latest version? Perhaps you could try it out in manifoldcad.org since you made that binding if I'm not mistaken.



What is your thought about my idea of using a linear extrude to simplify the warp code? Just for reference here, while trying to get my partial revolve working, I also wrote this code to try to do the same with a `revolve` instead of `extrude`. Note the necessity of using `Math.atan2` to figure out what 'slice' the vertex is in:

```js

const generateProfile = () => {
  const arcCenterPoint = [
    2.5,
    2.5
  ]
  const arcPoints = [...Array(7).keys()]
    .map(i => {
      const angle = i * Math.PI / (7)
      const x = Math.cos(angle) * 2.5
      const y = Math.sin(angle) * 2.5
      return [arcCenterPoint[0] + y, arcCenterPoint[1] - x]
    })

  const profile = [
    [0, 0],
    [2.5, 0],
    ...arcPoints,
    [2, 5],
    [2, 18],
    [0, 20]
  ]

  return profile
}

const profile = generateProfile()

const partialRevolve = (startAngle, endAngle, nSegments) => {
  const totalAngle = Math.abs(endAngle - startAngle)
  const angleStep = totalAngle / (nSegments - 1)

  const warpFunc = (vertex) => {
    // slice index is encoded z rotation
    const sliceAngle = Math.atan2(vertex[1], vertex[0]) + Math.PI
    const sliceRotStep = 2 * Math.PI / nSegments
    const zIndex = sliceAngle / sliceRotStep
    const finalAngle = zIndex * angleStep + startAngle

    const distance = Math.sqrt(vertex[0]*vertex[0] + vertex[1]*vertex[1])
    // do transform
    vertex[1] = distance * Math.sin(finalAngle)
    vertex[0] = distance * Math.cos(finalAngle)
  }

  return revolve(profile, nSegments)
    .warp(warpFunc)
}

const test = partialRevolve(-Math.PI / 4, Math.PI / 2, 6)

const result = compose([
  revolve(profile, 6)
  .translate([10, 0, 0]),
  test
])

```

It results in a partial revolve missing the cap slices:

![image](https://user-images.githubusercontent.com/66420755/233352550-61793cb3-06b5-4568-8670-809a6de42e8a.png)

If you feel inclined, perhaps you can use this code as a starting point to make a simple example of using the triangulate function to generate end caps!

**Reply by pca006132** [reply]
Weird, I looked at the code and found that we will not perform simplification when doing extrude. In fact, if you do a different transformation, e.g.
```js
    vertex[1] += Math.sin(angle)
    vertex[0] += Math.cos(angle)
```
it works fine:
![image](https://user-images.githubusercontent.com/12198657/233364044-43f7b4df-4d90-456f-9200-4b32298db28f.png)

I guess perhaps the warp function did something weird or later simplification messed it up?

**Reply by Loosetooth** [reply]
@pca006132 Wow, interesting. I can indeed reproduce this too. Just had a good look at the warp function, didn't see anything wrong with it. I think this points in the direction of broken simplification after doing the warp somehow.

Also, even with your transformation, I can still generate some strange triangles (seems like it skips a 'slice' somehow) by increasing the number of divisions:
```js
partialRevolve(-Math.PI / 4, Math.PI / 2, 10)
```
![image](https://user-images.githubusercontent.com/66420755/233373337-72b718b9-e0a0-45c3-b459-b5ee12bbb839.png)
![image](https://user-images.githubusercontent.com/66420755/233373450-22c9fced-d141-448e-8914-2628dff070c7.png)

**Reply by pca006132** [reply]
Yeah I guess this indicates other issues
Maybe we should try to simplify the example, e.g. simplify the polygon into simple shapes. Also, we can have a look if union will produce similar artifact. We should open an issue for this (but with a different title).

**Reply by elalish** [reply]
I bet I know what the problem is - let me try a fix. Regarding missing JS bindings, there are actually a bunch: https://github.com/elalish/manifold/issues/360; I'd love some help here if you have some spare cycles.

**Reply by elalish** [reply]
Sure enough, I was just missing two lines: #416. The key was noticing that the bug didn't repro if the `compose` was removed.

**Reply by Loosetooth** [reply]
I just tried with your fix, and I'm still having the problem. You're right that it only appears when either using `compose` or `union`. I made the example a bit smaller, and added all the different result calls in comments:

```js
const generateProfile = () => {
  const arcCenterPoint = [
    2.5,
    2.5
  ]

  const divisions = 10
  const arcPoints = [...Array(divisions).keys()]
    .map(i => {
      const angle = i * Math.PI / (divisions)
      const x = Math.cos(angle) * 2.5
      const y = Math.sin(angle) * 2.5
      return [arcCenterPoint[0] + y, arcCenterPoint[1] - x]
    })

  const profile = [
    [0, 0],
    ...arcPoints,
    [0, 2 * arcCenterPoint[1]]
  ]

  return profile
}

const profile = generateProfile()

const partialRevolve = (startAngle, endAngle, nSegments) => {
  const totalAngle = Math.abs(endAngle - startAngle)
  const angleStep = totalAngle / (nSegments - 1)

  const warpFunc = (vertex) => {
    // slice index is encoded z height
    const zIndex = nSegments - 1 - vertex[2]
    const angle = zIndex * angleStep + startAngle
    // do transform
    vertex[2] = vertex[1]
    vertex[1] = vertex[0] * Math.sin(angle)
    vertex[0] = vertex[0] * Math.cos(angle)
  }

  return extrude(profile, nSegments - 1, nSegments - 2)
    .warp(warpFunc)
}

const result = partialRevolve(-Math.PI / 4, Math.PI / 2, 17)
// const result = compose([partialRevolve(-Math.PI / 4, Math.PI / 2, 17)])
// const result = compose([
//   partialRevolve(-Math.PI / 4, Math.PI / 2, 10),
//   extrude(profile, 10 - 1, 10 - 2).translate([10, 0, 0])
// ])
// const result = union(
//   partialRevolve(-Math.PI / 4, Math.PI / 2, 10),
//   extrude(profile, 10 - 1, 10 - 2).translate([10, 0, 0])
// )
```

Only the first `result` line leads to the correct mesh. All the other ones still result in triangles being removed.

**Reply by elalish** [reply]
Okay, so first off, I'm guessing you're testing this in manifoldCAD.org? That's a bit tricky because it's a PWA - you end up having to reload it a couple of times to actually pull down any new changes. However, I just tried that and ran into a much worse problem - no output at all because some of the positions are NaN. So stay tuned while I look into that.

**Reply by elalish** [reply]
On the plus side, you can still use Save GLB, so you can inspect the output and see why it's invalid.

**Reply by Loosetooth** [reply]
Thanks, I'll try it again after force reloading.

If this might help you, I was just writing an equivalent test in cpp. Please, my CPP is bad, haven't used it in years. So it might have some other problems. However, it results in a segmentation fault on the `Manifold simplified = Manifold::Compose({shape});` line.

```cpp
TEST(Manifold, ComposeSimplification) {
  CrossSection circle =
      CrossSection::Circle(5, 20).Translate(glm::vec2(10.0, 10.0));

  Manifold shape = Manifold::Extrude(circle, 2, 10).Warp([](glm::vec3& v) {
    int nSegments = 10;
    double angleStep = 2.0 / 3.0 * M_PI / nSegments;
    int zIndex = nSegments - 1 - std::round(v.z);
    double angle = zIndex * angleStep;
    v.z = v.y;
    v.y = v.x * sin(angle);
    v.x = v.x * cos(angle);
  });

  auto propBefore = shape.GetProperties();

  Manifold simplified = Manifold::Compose({shape});
  auto propAfter = simplified.GetProperties();

  EXPECT_NEAR(propBefore.volume, propAfter.volume, 0.0001);
  EXPECT_NEAR(propBefore.surfaceArea, propAfter.surfaceArea, 0.0001);
  EXPECT_NEAR(propBefore.volume, 2, 0.0001);
  EXPECT_TRUE(true);
}
```

**Reply by Loosetooth** [reply]
Ah yes, just 'received' the new web version :)

Running the example code above with any of the lines with 'compose' or 'union' results in `RuntimeError: memory access out of bounds`. Which might correspond to the segmentation fault in the cpp test.

**Reply by pca006132** [reply]
@elalish it seems that you forgot to call finish after simplify topology. I added it and there is no segfault now.

**Reply by elalish** [reply]
Thank you, I was wondering if it was something like that. Okay, I'll go ahead and make a PR.

**Reply by elalish** [reply]
Okay, trying again with #418.

**Reply by jirihon** [reply]
@Loosetooth I will rather show you my `polyhedron` constructor that I use for sweeping and face shifting to showcase the `triangulate` function. It shows the whole process - how to triangulate and how to pack all the data for manifold.

You can ignore the code for vertex normals and uvs if you don't need them. I'm leaving it in TypeScript as it is more readable with types. It needs a little work to get it to run in manifoldcad.org, but I believe you can do it easily :+1: 

```typescript

/**
 * Make manifold from a set of polygonal faces
 *
 * @param vert_pos Vertex coordinates
 * @param face_polygons_vec Array of face polygons, each face can be described
 * 	by one or more polygons to allow for holes, polygon vertex is an index
 *  to the vert_pos array
 * @param face_normals Face normals, the face normal is needed only when the face polygon
 *  is not triangle or quadrilateral, allows to triangulate 3D polygons
 * @param vert_normals Optional vertex normals
 * @param vert_uvs Optional vertex uvs
 * @param merge Optional vertex merge map needed when vert_normals or vert_uvs are given
 * @param mf ManifoldStatic object
 * @returns Manifold object and triFace array mapping each triangle to the original face
 */
export function polyhedron(
	vert_pos: Vec3[],
	face_polygons_vec: number[][][],
	face_normals: Vec3[],
	vert_normals: Vec3[] | undefined,
	vert_uvs: Vec2[] | undefined,
	merge: Map<number, number> | undefined,
	mf: ManifoldStatic,
) {
	let numProp = 3;
	if (vert_normals !== undefined) {
		numProp += 3;
	}
	if (vert_uvs !== undefined) {
		numProp += 2;
	}
	const vertProperties = new Float32Array(vert_pos.length * numProp);
	for (const [i, vert] of vert_pos.entries()) {
		vertProperties[numProp * i] = vert[0];
		vertProperties[numProp * i + 1] = vert[1];
		vertProperties[numProp * i + 2] = vert[2];
	}
	if (vert_normals !== undefined) {
		for (const [i, normal] of vert_normals.entries()) {
			vertProperties[numProp * i + 3] = normal[0];
			vertProperties[numProp * i + 4] = normal[1];
			vertProperties[numProp * i + 5] = normal[2];
		}
	}
	if (vert_uvs !== undefined) {
		for (const [i, uv] of vert_uvs.entries()) {
			vertProperties[numProp * i + numProp - 2] = uv[0];
			vertProperties[numProp * i + numProp - 1] = uv[1];
		}
	}
	const tri_verts: Vec3[] = [];
	const triFace: number[] = [];

	for (const [i, polygons] of face_polygons_vec.entries()) {
		let tris: Vec3[] = [];
		if (polygons.length === 1 && polygons[0].length === 3) {
			tris.push(polygons[0] as Vec3);
		} else if (polygons.length === 1 && polygons[0].length === 4) {
			const p = polygons[0];
			tris.push([p[0], p[1], p[3]]);
			tris.push([p[1], p[2], p[3]]);
		} else {
			const projection = axis_aligned_projection(face_normals[i]);
			let idx = 0;
			const vert_i_map: number[] = [];
			const polygons_2d = polygons.map((polygon) =>
				polygon.map((vert_i) => {
					vert_i_map[idx++] = vert_i;
					return projection(vert_pos[vert_i]);
				}),
			);
			const result = mf.triangulate(polygons_2d);
			tris = result.map((tri) => tri.map((vert_i) => vert_i_map[vert_i]) as Vec3);
		}
		for (const tri of tris) {
			tri_verts.push(tri);
			triFace.push(i);
		}
	}
	const triVerts = new Uint32Array(tri_verts.length * 3);
	for (const [i, tri] of tri_verts.entries()) {
		triVerts[3 * i] = tri[0];
		triVerts[3 * i + 1] = tri[1];
		triVerts[3 * i + 2] = tri[2];
	}
	let mergeFromVert: Uint32Array | undefined;
	let mergeToVert: Uint32Array | undefined;
	if (merge !== undefined) {
		mergeFromVert = new Uint32Array(merge.size);
		mergeToVert = new Uint32Array(merge.size);
		let i = 0;
		for (const [from, to] of merge.entries()) {
			mergeFromVert[i] = from;
			mergeToVert[i] = to;
			i++;
		}
	}
	const mesh_options = {
		numProp,
		vertProperties,
		triVerts,
		mergeFromVert,
		mergeToVert,
	};
	const mesh = new mf.Mesh(mesh_options);
	const manifold = new mf.Manifold(mesh);
	return {
		manifold,
		triFace,
	};
}

export function axis_aligned_projection(normal: Vec3): (v: Vec3) => Vec2 {
	const abs_normal: Vec3 = [Math.abs(normal[0]), Math.abs(normal[1]), Math.abs(normal[2])];
	let z: number;
	let x: number;
	let y: number;

	if (abs_normal[2] > abs_normal[0] && abs_normal[2] > abs_normal[1]) {
		// x-y plane
		z = normal[2];
		x = 0;
		y = 1;
	} else if (abs_normal[1] > abs_normal[0]) {
		// z-x plane
		z = normal[1];
		x = 2;
		y = 0;
	} else {
		// y-z plane
		z = normal[0];
		x = 1;
		y = 2;
	}
	if (z < 0) {
		return (v) => [v[y], v[x]];
	} else {
		return (v) => [v[x], v[y]];
	}
}
```

**Comment by Loosetooth** [comment, marked as answer]
> Okay, trying again with https://github.com/elalish/manifold/pull/418.

After this pull request it works as expected! Thanks for much for everyone's time and effort!
